### PR TITLE
Fixes how annotations are fetched from annotated types

### LINKED ISSUE #2603: `st.builds()` works incorrectly with annotated `__init__`
The simplest reproduction is:

```python
from typing import Generic, TypeVar

from hypothesis import given
from hypothesis import strategies as st

_ValueType = TypeVar('_ValueType')


class Wrapper(Generic[_ValueType]):
    _inner_value: _ValueType

    def __init__(self, inner_value: _ValueType) -> None:
        self._inner_value = inner_value


@given(st.builds(Wrapper))
def test_wrapper(wrapper):
    assert isinstance(wrapper, Wrapper)

```

Outputs:

```
====================================== FAILURES ======================================
____________________________________ test_wrapper ____________________________________

    @given(st.builds(Wrapper))
>   def test_wrapper(wrapper):

ex2.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.7/site-packages/hypothesis/core.py:684: in _execute_once_for_engine
    escalate_hypothesis_internal_error()
.venv/lib/python3.7/site-packages/hypothesis/core.py:655: in _execute_once_for_engine
    result = self.execute_once(data)
.venv/lib/python3.7/site-packages/hypothesis/core.py:610: in execute_once
    result = self.test_runner(data, run)
.venv/lib/python3.7/site-packages/hypothesis/executors.py:52: in default_new_style_executor
    return function(data)
.venv/lib/python3.7/site-packages/hypothesis/core.py:550: in run
    args, kwargs = data.draw(self.search_strategy)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:889: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/collections.py:57: in do_draw
    return tuple(data.draw(e) for e in self.element_strategies)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/collections.py:57: in <genexpr>
    return tuple(data.draw(e) for e in self.element_strategies)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:650: in do_draw
    result = self.pack(data.draw(self.mapped_strategy))
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/lazy.py:150: in do_draw
    return data.draw(self.wrapped_strategy)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:650: in do_draw
    result = self.pack(data.draw(self.mapped_strategy))
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/collections.py:57: in do_draw
    return tuple(data.draw(e) for e in self.element_strategies)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/collections.py:57: in <genexpr>
    return tuple(data.draw(e) for e in self.element_strategies)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/lazy.py:150: in do_draw
    return data.draw(self.wrapped_strategy)
.venv/lib/python3.7/site-packages/hypothesis/internal/conjecture/data.py:884: in draw
    return strategy.do_draw(self)
.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/strategies.py:650: in do_draw
    result = self.pack(data.draw(self.mapped_strategy))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = ((), {})

>       lambda value: target(*value[0], **value[1])  # type: ignore
    )
E   TypeError: __init__() missing 1 required positional argument: 'inner_value'

.venv/lib/python3.7/site-packages/hypothesis/strategies/_internal/core.py:1313: TypeError
```

Why does this happen?

My debugging session shows that:
- `required` [here](https://github.com/HypothesisWorks/hypothesis/blob/d1efdd688d889931427afa0defdc4abdbfc95c1a/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L1282) is equal to `{'inner_value'}`
- But, since [here we fetch annotations](https://github.com/HypothesisWorks/hypothesis/blob/d1efdd688d889931427afa0defdc4abdbfc95c1a/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L1291) from the object itself, not its `__init__` method. It results in `{'_inner_value': ~_ValueType}`
- And since these two sets do not match, the [resulting set here](https://github.com/HypothesisWorks/hypothesis/blob/d1efdd688d889931427afa0defdc4abdbfc95c1a/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L1297) is empty
- And that's why we have this error 😢 

I can fix it! The only question I have is: 
Is it ok to chage how we get annotation from `target` if we are dealing with `__init__`?

### PR DESCRIPTION
**Pull request by sobolevn**
Closes #2603 

@Zac-HD I have followed your idea. It seems to be the easiest way. Hacky, but working! 👍

### CODE REVIEW

**Comment by Zac-HD** [comment]
If you're using the update method we'll need to reverse the order in which we attempt them, because update overrides earlier keys.

You'll also hit SyntaxError on Python 3.5 - this test will have to go in a file which is not collected under py35 (check the conf.py), or just wait a few days and I'll drop py35 entirely - EOL is at the end of the month.

**Comment by sobolevn** [comment]
> If you're using the update method we'll need to reverse the order in which we attempt them, because update overrides earlier keys.

Shouldn't `__init__` annotations be more specific and override any class-level ones?

Imagine: 

```python
class Test(Generic[T]):
   value: T

   def __init__(self, value: Optional[T]) -> None:
       if value is None:
           value = default_of_t()
       ...
```

If we first use class-level annotation and the override them with `__init__` annotations, we would cover both branches.
But, if class-level annotations are more important, then it would not be correct.

The same about any other type-casing in costructor. 
That's why I have put them in the last order.

**Comment by sobolevn** [comment]
We can even add an explicit test for it:

```python
class AnnotatedConstructor(Generic[_ValueType]):
    value: _ValueType

    def __init__(self, value: int) -> None:
         assert isinstance(value, int)
```

**Comment by Zac-HD** [comment]
Makes sense! If the `__signature__` attribute is defined it should probably take highest priority then.

(eill review properly tomorrow)

File: hypothesis-python/RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+RELEASE_TYPE: patch
+
+This patch fixes a problem with ``st.builds()`` that was not able to
```
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst:3]:
```suggestion
This patch fixes a problem with :func:`~hypothesis.strategies.builds` that was not able to
```

File: hypothesis-python/src/hypothesis/internal/compat.py [hunk 2]
```diff
@@ -136,8 +136,20 @@ def get_type_hints(thing):
         NameError for unresolvable forward references, just return an empty dict.
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/compat.py:136]: Please also update the docstring of this function to describe what it does now, and why we need the additional logic for classes.

File: hypothesis-python/tests/cover/test_lookup_py36.py [hunk 3]
```diff
@@ -31,3 +32,53 @@ class TreeForwardRefs(typing.NamedTuple):
...
+    __signature__ = signature(use_signature)
+
+    def __init__(self, value: int) -> None:
+        """By this example we show, that ``__signature__`` is the most important source."""
+        assert isinstance(value, str)
+
+
+@given(st.data())
+def test_signature_is_the_most_important_source(data):
+    """Signature types should take presence over all other annotations."""
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_lookup_py36.py:83]:
```suggestion
    """Signature types should take precedence over all other annotations."""
```

**Review by Zac-HD — COMMENTED** [review]
This looks fantastic - thanks @sobolevn :heart_eyes: 

Some small comments below on cross-references and docstrings, but I'm very happy with the substance of the patch now - and the tests are great!

File: hypothesis-python/tests/cover/test_lookup_py36.py [hunk 4]
```diff
@@ -31,3 +32,53 @@ class TreeForwardRefs(typing.NamedTuple):
...
+    value: _ValueType  # the same name we have in `__init__`
+
+    def __init__(self, value: int) -> None:
+        """By this example we show, that ``int`` is more important than ``_ValueType``."""
+        assert isinstance(value, int)
+
+
+@given(st.data())
+def test_constructor_is_more_important(data):
+    """Constructor types should take presence over all other annotations."""
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_lookup_py36.py:63]: Oops, I missed this one in review :sweat_smile:

Comment (sobolevn) [reply hypothesis-python/tests/cover/test_lookup_py36.py:63]: 👍

**Review by Zac-HD — APPROVED** [review]
:tada:

**Comment by sobolevn** [comment]
Thanks a lot for your help!

**Comment by Zac-HD** [comment]
Always a pleasure :smiling_face_with_three_hearts:

**Comment by sobolevn** [comment]
With fixing this issue, we are ready to release new version of `returns`! Check out this beauty: https://github.com/dry-python/returns/blob/master/tests/test_contrib/test_hypothesis/test_laws/test_custom_type_with_init.py

Now users can write their own container with fully typed method (due to HKT support) and enforced laws (due to hypothesis)! 🤩

**Pull request merged**